src/components/ui/Panel.jsx

```diff
@@ -124,60 +152,87 @@ class Panel extends React.Component {
   */
   move = event => {
     const clientY = getEventClientY(event);
-    const currentHeight = this.startHeight + (this.startClientY - clientY);
-
-    if (this.props.size === 'maximized' &&
-        this.panelContentRef.current.scrollTop > 0) {
-      /* User is scrolling inside the panel content,
-         update startClientY to ignore current swipe gesture */
-      this.startClientY = clientY;
+    const visibleHeight = Math.ceil(window.innerHeight - clientY + this.startClientYOffset);
+    const { scrollTop } = this.panelContentRef.current;
+
+    if (this.state.translateY === 0 &&
+        this.props.size === 'maximized' &&
+        scrollTop === 0 &&
+        visibleHeight >= this.startHeight) {
+      /* User is starting to scroll content area from bottom to top
+       * Do not prevent default */
       return;
     }
 
-    if (this.props.size === 'maximized' &&
-        this.panelContentRef.current.scrollTop === 0 &&
-        currentHeight >= this.state.currentHeight) {
-      // User is starting to scroll content area from bottom to top, do nothing
+    if (this.state.translateY === 0 &&
+        this.props.size === 'maximized' && scrollTop > 0) {
+      /* User is already scrolling inside the panel content.
+       * Update startClientY to ignore current swipe gesture */
+      const rect = this.panelDOMElement.getBoundingClientRect();
+      this.startClientY = clientY;
+      this.stopHeight = window.innerHeight - rect.top;
+      this.startClientYOffset = clientY - rect.top;
       return;
     }
 
-    this.setState({ currentHeight, holding: true });
+    event.preventDefault();
+
+    const translateY = visibleHeight >= this.state.height
+      ? 0 // Prevent panel to be moved above the top bar
+      : this.state.height - visibleHeight;
+
+    this.setState({ translateY });
+  }
+
+  getTranslateY(size) {
+    const { fitContent } = this.props;
+    const { height } = this.state;
+    const panelHeight = this.panelContentRef.current.offsetHeight;
+    const values = {
+      'default': height - (fitContent
+        ? panelHeight + FIT_CONTENT_PADDING
+        : DEFAULT_SIZE),
+      'minimized': height - (fitContent
+        ? panelHeight + FIT_CONTENT_PADDING
+        : DEFAULT_MINIMIZED_SIZE),
+      'maximized': 0,
+    };
+
+    return values[size];
   }
 
   /**
    * Triggered on mouse up of the panel resizer
    * @param {MouseEvent|TouchEvent} event
    */
-  stopResize = event => {
+  stopResize = _ => {
     this.removeListeners();
-    const clientY = getEventClientY(event);
-    const currentHeight = this.startHeight + (this.startClientY - clientY);
+    const rect = this.panelDOMElement.getBoundingClientRect();
```

Comment:
These two lines are done multiple times, I wonder if it's worth putting them in a function.

Reply:
it seems there is only `this.panelDOMElement.getBoundingClientRect()` which is common at 3 different lines. In these 3 cases, the computations after getting `getBoundingClientRect` are either for computations relates to `start` or `stop`, so it does not seem to share any code. Do you think we could still wrap some lines into a function ?

Reply:
Right, let's forget it :) We are probably going to come back to this component after, so if some simplification is obvious we'll do it.